irc3/plugins/slack.py

```diff
@@ -0,0 +1,247 @@
+
+Install aiohttp::
+
+    $ pip install aiohttp
+
+Usage
+=====
+
+Create a bridge between slack and irc
+
+..
+    >>> from irc3.testing import IrcBot
+    >>> from irc3.testing import ini2config
+    >>> config = ini2config("""
+    ... [bot]
+    ... includes =
+    ...     irc3.plugins.slack
+    ... [irc3.plugins.slack]
+    ... token = xoxb-slackbottoken
+    ... [irc3.plugins.slack.channels]
+    ... slack =
+    ...     ${#}irc
+    ... """)
+    >>> bot = IrcBot(**config)
+
+.. note::
+
+    Be sure to invite the bot in slack to the channels it should be bridging
+
+'''
+
+EMOJIS = {
+  "smile": ":)",
+  "simple_smile": ":)",
+  "smiley": ":-)",
+  "grin": ":D",
+  "wink": ";)",
+  "smirk": ";)",
+  "blush": ":$",
+  "stuck_out_tongue": ":P",
+  "stuck_out_tongue_winking_eye": ";P",
+  "stuck_out_tongue_closed_eyes": "xP",
+  "disappointed": ":(",
+  "astonished": ":O",
+  "open_mouth": ":O",
+  "heart": "<3",
+  "broken_heart": "</3",
+  "confused": ":S",
+  "angry": ">:(",
+  "cry": ":,(",
+  "frowning": ":(",
+  "imp": "]:(",
+  "innocent": "o:)",
+  "joy": ":,)",
+  "kissing": ":*",
+  "laughing": "x)",
+  "neutral_face": ":|",
+  "no_mouth": ":-",
+  "rage": ":@",
+  "smiling_imp": "]:)",
+  "sob": ":,'(",
+  "sunglasses": "8)",
+  "sweat": ",:(",
+  "sweat_smile": ",:)",
+  "unamused": ":$"
+}
+
+
+@irc3.plugin
+class Slack:
+
+    def __init__(self, bot):
+        self.bot = bot
+        self.config = self.bot.config.get(__name__, {})
+        self.channels = self.bot.config.get(
+            '{0}.channels'.format(__name__), {}
+        )
+        self.clean_channels()  # remove has values from channels
+        autojoins = set().union(*self.channels.values())
+        self.bot.log.debug('Adding to autojoins list: {autojoins}')
+        self.bot.config.setdefault('autojoins', []).extend(autojoins)
+        self.slack_channels = {}
+        self.slack_users = {}
+        if 'token' not in self.config:
+            self.bot.log.warning('No slack token is set.')
+
+    def clean_channels(self):
+        self.channels.pop('#', None)
+        self.channels.pop('hash', None)
+
+    async def api_call(self, method, data=None):
+        """Slack API call."""
+        async with aiohttp.ClientSession() as session:
+            form = aiohttp.FormData(data or {})
+            form.add_field('token', self.config['token'])
+            async with session.post(
+                'https://slack.com/api/{0}'.format(method), data=form
+                    ) as response:
+                if response.status != 200:
+                    raise Exception(
+                        '{0} with {1} failed.'.format(method, data)
+                    )
+                return await response.json()
+
+    def server_ready(self):
+        irc3.asyncio.ensure_future(self.connect())
+        self.bot.log.debug('Listening to Slack')
+
+    def parse_text(self, message):
+
+        def getChannelById(matchobj):
```

Comment:
Why do you need closures? Also, why camel case ?

Reply:
Would just using a lambda be better?

Reply:
You can use standard methods. Or functions with functools.partial. Also your matches array can be defined in init to avoid redefining it for each calls:
```
self.matches = [
    (r'', self.get_emojis)
]
``` 

Or with functions:

```
self.matches = [
    (r'', partial(get_emojis, self))
]
``` 

No need for closure in both cases